chemprop/models/ffn.py

```diff
@@ -0,0 +1,355 @@
+        :param ffn_base: The shared base layers (all but the last) of the FFN between tasks.
+        :param constraint: Whether to apply constraint to output.
+        :param ffn_type: The type of target (atom or bond).
+        :param shared_ffn: Whether to share weights in the ffn between different atom tasks and bond tasks.
+        :param weights_ffn_num_layers: Number of layers in FFN for determining weights used to correct the constrained targets.
+        """
+        super().__init__()
+
+        if num_layers == 1:
+            base_output_size = features_size
+        else:
+            base_output_size = hidden_size
+
+        if constraint:
+            if shared_ffn:
+                self.ffn = ffn_base
+            else:
+                if num_layers > 1:
+                    self.ffn = nn.Sequential(
+                        build_ffn(
+                            first_linear_dim=features_size,
+                            hidden_size=hidden_size,
+                            num_layers=num_layers - 1,
+                            output_size=hidden_size,
+                            dropout=dropout,
+                            activation=activation,
+                        ),
+                        activation,
+                    )
+                else:
+                    self.ffn = nn.Identity()
+            self.ffn_readout = build_ffn(
+                first_linear_dim=base_output_size,
+                hidden_size=hidden_size,
+                num_layers=1,
+                output_size=output_size,
+                dropout=dropout,
+                activation=activation,
+            )
+            self.weights_readout = build_ffn(
+                first_linear_dim=base_output_size,
+                hidden_size=hidden_size,
+                output_size=1,
+                num_layers=weights_ffn_num_layers,
+                dropout=dropout,
+                activation=activation,
+            )
+        else:
+            if shared_ffn:
+                self.ffn_readout = nn.Sequential(
+                    ffn_base,
+                    build_ffn(
+                        first_linear_dim=base_output_size,
+                        hidden_size=hidden_size,
+                        num_layers=1,
+                        output_size=output_size,
+                        dropout=dropout,
+                        activation=activation,
+                    ),
+                )
+            else:
+                self.ffn_readout = build_ffn(
+                    first_linear_dim=features_size,
+                    hidden_size=hidden_size,
+                    num_layers=num_layers,
+                    output_size=output_size,
+                    dropout=dropout,
+                    activation=activation,
+                )
+        self.constraint = constraint
+        self.ffn_type = ffn_type
+
+    def forward(
+        self,
+        input: Tuple[torch.tensor, List, torch.tensor, List, torch.tensor],
```

Comment:
`torch.tensor` is a function not a class

Reply:
fixed